### ISSUE TITLE
Error after failing payment in account/orders

### ISSUE THREAD

**Issue by igor-tv** [OFFLINE-GmbH/oc-mall-plugin#360]
After failing to pay for an order using PayPal, I tried to go to the account's order page and received this error:
```
An exception has been thrown during the rendering of a template ("Trying to get property 'latest_file' of non-object").
/Applications/MAMP/htdocs/shop/plugins/offline/mall/components/orderslist/productlist.htm line 32
```
In this order there were only non-virtual goods.

**Comment by igor-tv** [comment]
Sorry, I was unable to replay this behavior. Perhaps the problem was in the wrong settings on my part. I changed the currency code.

**Comment by igor-tv** [comment]
But I found another minor bug. If you have an order with a failed payment, and for some reason you changed the currency code on the backend, then you will not be able to go to the details page of this unsuccessful order on the backend.

Error:
```
No query results for model [OFFLINE \ Mall \ Models \ Currency].
/Applications/MAMP/htdocs/shop/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php line 414
```

**Comment by tobias-kuendig** [comment]
This is fixed in df69e4e, thanks @igor-tv!

**Comment by maartenmachiels** [comment]
Hi guys,

I was able to reproduce the error... After loggin in in a users account, I get the following stack trace. Do you know what the issue might be?

<html>
<body>
<h1 style="box-sizing: border-box; font-size: 50px; margin: 20px 0px 15px; font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-weight: 300; line-height: 1.1; color: rgb(64, 82, 97); text-transform: uppercase; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(243, 243, 243); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span> </span>ERROR</h1><p class="lead" style="box-sizing: border-box; margin: 0px 0px 20px; font-size: 16px; font-weight: 300; line-height: 1.4; color: rgb(64, 82, 97); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(243, 243, 243); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">We're sorry, but an unhandled error occurred. Please see the details below.</p><div class="exception-name-block" style="box-sizing: border-box; margin-bottom: 15px; color: rgb(64, 82, 97); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(243, 243, 243); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div style="box-sizing: border-box; font-weight: 300; padding: 10px 15px; background: rgb(200, 69, 48); color: white; position: relative;">An exception has been thrown during the rendering of a template ("Trying to get property 'latest_file' of non-object").</div><p style="box-sizing: border-box; margin: 0px; font-weight: 300; padding: 10px 15px; background: rgb(235, 235, 235);">/var/www/vhosts/brightlab2021.chocolata.be/httpdocs/plugins/offline/mall/components/orderslist/productlist.htm<span> </span><span style="box-sizing: border-box; color: rgb(137, 159, 177);">line</span><span> </span>32</p></div><ul class="indicators" style="box-sizing: border-box; margin: 0px; list-style: none; padding: 0px; color: rgb(64, 82, 97); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(243, 243, 243); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box; float: left; margin-right: 50px;"><h3 style="box-sizing: border-box; font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-weight: 300; line-height: 12px; color: inherit; margin-top: 20px; margin-bottom: 7px; font-size: 12px; text-transform: uppercase;">TYPE</h3><p style="box-sizing: border-box; margin: 0px 0px 10px; font-size: 20px; font-weight: 600; overflow-wrap: break-word;">Twig Template</p></li><li style="box-sizing: border-box; float: left; margin-right: 0px;"><h3 style="box-sizing: border-box; font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-weight: 300; line-height: 12px; color: inherit; margin-top: 20px; margin-bottom: 7px; font-size: 12px; text-transform: uppercase;">EXCEPTION</h3><p style="box-sizing: border-box; margin: 0px 0px 10px; font-size: 20px; font-weight: 600; overflow-wrap: break-word;">Cms\Classes\CmsException</p></li></ul><div style="box-sizing: border-box; color: rgb(64, 82, 97); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(243, 243, 243); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div id="highlighter_485608" class="syntaxhighlighter  htmlscript" style="box-sizing: border-box; margin-bottom: 30px; background: white; font-style: normal !important; font-variant: normal !important; font-weight: normal !important; font-stretch: normal !important; font-size: 12px !important; line-height: 20.4px; font-family: Monaco, Menlo, &quot;Ubuntu Mono&quot;, &quot;Droid Sans Mono&quot;, &quot;Courier New&quot;, monospace !important;">

26272829303132333435363738 | {% endif %}{% if product.product_file_grants.count > 0 %}    <br/>    {% for grant in product.product_file_grants %}        <br/>        <a href="{{ grant.download_link }}">            Download {{ grant.display_name }}            {{ loop.length > 1 ? ('#' ~ loop.index) }}        </a>        {% if grant.expires_at %}            <br>valid until {{ grant.expires_at.toDateString() }}        {% endif %}        {% if grant.max_download_count %}
-- | --



</body>
</html>

**Comment by maartenmachiels** [comment]
A quick update - I managed to get rid of the error. The database table `offline_mall_product_files` was empty, but the table `offline_mall_product_file_grants` did have a lot of references. I'm guessing the `offline_mall_product_file_grants` tried to look up files that don't exist. What do you think?

Is it the case that when a product file is deleted, the related file grant is deleted aswel? If not, I'm guessing that is the source of the error...

**Comment by maartenmachiels** [comment]
Hi, I'm doing some more testing and it does seem like even though a product has no associated files, the table offline_mall_product_file grants is being populated. The problem is that is then displays this strange message: "Download download 114". If you click on it, the file is not found, because it doesn't exist. 

What can we do to tackle this?

![image](https://user-images.githubusercontent.com/3584586/136002357-57bcf301-0dc4-4428-9e74-74bf59f7e0ef.png)

**Comment by tobias-kuendig** [comment]
The log message above comes from here:
https://github.com/OFFLINE-GmbH/oc-mall-plugin/blob/develop/classes/downloads/VirtualProductFileDownload.php#L65

This method is only triggered when you visit the download link. Where do you list the download link? If you don't have any file attached, why show a download link?

Do you generate a file on your own?

**Comment by maartenmachiels** [comment]
Hi @tobias-kuendig 

Thanks so much for your answer. You are correct - I don't want to show a download link if there is no file associated with the virtual product. But at the moment the download link is being displayed regardless:

The download link is displayed in the following partials:

`/plugins/offline/mall/components/orderslist/productslist.htm`

```
                {% if product.product_file_grants.count > 0 %}
                    <br/>
                    {% for grant in product.product_file_grants %}
                        <br/>
                        <a href="{{ grant.download_link }}">
                            Download {{ grant.display_name }}
                            {{ loop.length > 1 ? ('#' ~ loop.index) }}
                        </a>
                        {% if grant.expires_at %}
                            <br>valid until {{ grant.expires_at.toDateString() }}
                        {% endif %}
                        {% if grant.max_download_count %}
                            <br>max. {{ grant.max_download_count }} downloads
                        {% endif %}
                    {% endfor %}
                {% endif %}
```

And also in one of the fragments of one of the emails sent by the system `mall.order.table`

So basically - if you create a virtual product and do not link a file to it, the above code does still display a download link in the format `Download Download ID`.

I assume that if there is no file connected to a virtual product, a file grant should not be created. Am I overlooking something?

Thanks for looking into it, I appreciate it. 

P.S.: My client is absolutely thrilled with your e-commerce solution. It is so incredibly complete and practical right out of the box. And it's a joy to work with. We only need to remove a few small hiccups and we're golden!

**Comment by tobias-kuendig** [comment]
> I assume that if there is no file connected to a virtual product, a file grant should not be created. Am I overlooking something?

This is correct. Before this line, you can try something like this (untested):

https://github.com/OFFLINE-GmbH/oc-mall-plugin/blob/b104849606639f2929f676c189d52a16aefc8628/models/ProductFileGrant.php#L64

```php
// Do not create a download grant if the product has no files attached.
if (!$orderProduct->product->latest_file) {
    return;
}

```

This should get rid of the empty download grants, and since no grants are created, the code in the partial you linked above should be skipped  (`{% if product.product_file_grants.count > 0 %}`)

If this fixes your issue a PR is as always welcome!

**Comment by maartenmachiels** [comment]
Hi @tobias-kuendig,

You're the master!

Just tested it out, both with and without a file attached to the product. Now it works perfectly. I've created a pull request here:
[https://github.com/OFFLINE-GmbH/oc-mall-plugin/pull/862](https://github.com/OFFLINE-GmbH/oc-mall-plugin/pull/862)

Please note one small thing, not important for me on the short term: I notice that the download link is available even if the order hasn't been paid yet. I can easily circumvent this by changing the Twig code, but I just wanted to let you know.

Thanks for all of your work and help.

**Comment by dathwa** [comment]
This is causing serious issues with my virtual downloads.
My products do NOT have a file associated. A pdf is generated on the file, after the `mall.product.file_grant.created` event is triggered.
But these changes ([https://github.com/OFFLINE-GmbH/oc-mall-plugin/pull/862/commits/52f112c5d36b091b4a082472590cd407455c6121](url)) are preventing this.
I've commented out for my purposes, but can we have a fix for this?
